monix-reactive/jvm/src/main/scala/monix/reactive/compress/package.scala

```diff
@@ -0,0 +1,172 @@
+ * limitations under the License.
+ */
+
+package monix.reactive
+
+import java.time.Instant
+import java.util.zip.Deflater
+
+import monix.reactive.compress.internal.operators.{DeflateOperator, GunzipOperator, GzipOperator, InflateOperator}
+
+import scala.collection.compat.immutable.ArraySeq
+
+// From https://github.com/typelevel/fs2/blob/main/core/jvm/src/main/scala/fs2/compression.scala
+package object compress {
+  private[compress] val zeroByte: Byte = 0
+  private[compress] val gzipMagicFirstByte: Byte = 0x1f.toByte
+  private[compress] val gzipMagicSecondByte: Byte = 0x8b.toByte
+
+  private[compress] object gzipFlag {
+    def apply(flags: Byte, flag: Byte): Boolean = (flags & flag) == flag
+
+    def apply(flags: Byte, flag: Int): Boolean = (flags & flag) == flag
+
+    def ftext(flags: Byte): Boolean = apply(flags, FTEXT)
+
+    def fhcrc(flags: Byte): Boolean = apply(flags, FHCRC)
+
+    def fextra(flags: Byte): Boolean = apply(flags, FEXTRA)
+
+    def fname(flags: Byte): Boolean = apply(flags, FNAME)
+
+    def fcomment(flags: Byte): Boolean = apply(flags, FCOMMENT)
+
+    def reserved5(flags: Byte): Boolean = apply(flags, RESERVED_BIT_5)
+
+    def reserved6(flags: Byte): Boolean = apply(flags, RESERVED_BIT_6)
+
+    def reserved7(flags: Byte): Boolean = apply(flags, RESERVED_BIT_7)
+
+    val FTEXT: Byte = 1
+    val FHCRC: Byte = 2
+    val FEXTRA: Byte = 4
+    val FNAME: Byte = 8
+    val FCOMMENT: Byte = 16
+    val RESERVED_BIT_5 = 32
+    val RESERVED_BIT_6 = 64
+    val RESERVED_BIT_7: Int = 128
+  }
+
+  private[compress] object gzipExtraFlag {
+    val DEFLATE_MAX_COMPRESSION_SLOWEST_ALGO: Byte = 2
+    val DEFLATE_FASTEST_ALGO: Byte = 4
+  }
+
+  private[compress] object gzipCompressionMethod {
+    val DEFLATE: Byte = Deflater.DEFLATED.toByte
+  }
+
+  implicit class RichByteObservable(source: Observable[Byte]) {
+
+    /** Returns a stream that incrementally compresses input into the GZIP format
+      * Internally it pages stream into chunks, where each incoming chunk is compressed at once, so it can utilize thread for long time if chunks are big.
+      *
+      * @param bufferSize             Size of buffer used internally, affects performance.
+      * @param chunkSize              Size of chunks, for most of the use-cases should be equal to bufferSize
+      * @param compressionParameters  see [[CompressionParameters]]
+      * @param fileName               optional file name
+      * @param comment                optional file modification time
+      * @param modificationTime       optional file comment
+      * @return
+      */
+    def gzip(
+      bufferSize: Int = 32 * 1024,
+      chunkSize: Int = 32 * 1024,
+      compressionParameters: CompressionParameters = CompressionParameters.Default,
+      fileName: Option[String] = None,
+      comment: Option[String] = None,
+      modificationTime: Option[Instant] = None
+    ): Observable[Byte] = {
```

Comment:
Is it desirable to return `Observable[Byte]` ?

I feel like it would be better to take and return `Observable[Array[Byte]]` and if the users need `Observable[Byte]` then they could do buffering and concatenation themselves.

fs2 and ZIO return Stream[Byte] but in reality, it's a stream of Chunk[Byte] which has to be explicit in `Observable`

Reply:
When you read data in a streaming fashion in binary format (e.g. from file) you mostly get `Observable[Byte]`, so that would require users to always do buffering before using any compression transformer. Also they would need to know what chunk size to choose. Choosing too small or too big value can cause performance degradation. This explains my intention behind using `Observable[Byte]` as the input.

To not introduce any confusion and be transparent for the end users I decided to do not change the type and return `Observable[Byte]`

Reply:
I was wrong, you might get `Stream[Byte]` but that's rather popular among libraries that chunk data under the hood (fs2/zio) which effectively means that you always deal with `Stream[Array[Byte]]` either explicitly or not. That's good as transforming bytes one by one might not be the optimal solution in terms of performance.

Reply:
Also monix-nio returns `Observable[Array[Byte]]` and if we should be compatible with anything, that should be monix-nio.

Reply:
As another reason for `Array`, Akka Streams (which doesn't do chunking) uses `ByteString` which extends `IndexedSeq[Byte]`

> When you read data in a streaming fashion in binary format (e.g. from file) you mostly get Observable[Byte], so that would require users to always do buffering before using any compression transformer.

Related operators, like `Observable.fromInputStream` return `Observable[Array[Byte]]` and if we add more, I feel like we should emit `Array[Byte]` for performance reasons, as you mentioned.

>  Also they would need to know what chunk size to choose. 

That's a fair point, maybe we could add a comment with a suggestion/example how to use it with `Observable[Byte]`?

Reply:
Ideally we would write a small benchmark on both to compare their performance.

Reply:
About that:
> That's a fair point, maybe we could add a comment with a suggestion/example how to use it with Observable[Byte]?

I thought that maybe such helpful extension method could be declared for those who will wonder?

  ```
final def chunked[B >: A](chunkSize :Int = 8 * 1024)(implicit ev: B =:= Byte) : Observable[Array[Byte]] = {
    bufferTumbling(chunkSize)
      .map(seq => ev.liftCo[Seq](seq).toArray)
  }
```

To make it more discover-able I would define it on `Observable` itself although that's a precedence.

Reply:
I think this operator might make sense for performance scenarios, not only for `Byte`.

`bufferTumbling` uses `Array` as a buffer but it converts it to `Seq` before sending downstream.
We could copy-paste implementation and send the copy of the array without the conversion.

Although it would be better to do with benchmarks (there are some that use bufferTumbling and it's a bottleneck) so I feel like it would be preferable to tackle in another PR